billing/loader.go

```diff
@@ -0,0 +1,236 @@
+package billing
+
+import (
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"regexp"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/twinj/uuid"
+)
+
+type Loader struct {
+	reportName  string
+	report      *billingReport
+	concurrency int
+	wg          sync.WaitGroup
+	json        []byte
+	injector    Injector
+}
+
+func NewLoader(injector Injector, concurrency int) *Loader {
+	return &Loader{
+		concurrency: concurrency,
+		wg:          sync.WaitGroup{},
+		json:        []byte{},
+		injector:    injector,
+	}
+}
+
+func (l *Loader) ProcessFile(reportName string, billingFile string) {
+	l.reportName = reportName
+	l.report = l.openBillingReport(billingFile)
+
+	valuesSink := make(chan []string)
+	reportSinks := make([]chan *billingRecord, 0, 0)
+
+	for i := 0; i < runtime.GOMAXPROCS(l.concurrency); i++ {
+		reportSink := make(chan *billingRecord)
+		reportSinks = append(reportSinks, reportSink)
+		l.wg.Add(2)
+		go l.parseRecord(valuesSink, reportSink, l.report)
+		go l.saveRecord(reportSink)
+	}
+
+	for {
+		values, err := l.report.csvReader.Read()
+		if err == io.EOF {
+			break
+		}
+		valuesSink <- values
+	}
+	close(valuesSink)
+	l.wg.Wait()
+}
+
+type fieldMapper map[string]map[interface{}]map[string]interface{}
```

Comment:
This type has been removed, because I managed to worked around it previously with reflection; I forgot to remove it though :D - so no need for it!